radixdlt-core/radixdlt/src/main/java/com/radixdlt/api/service/reducer/AllValidators.java

```diff
@@ -0,0 +1,265 @@
+/*
+ * (C) Copyright 2021 Radix DLT Ltd
+ *
+ * Radix DLT Ltd licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except in
+ * compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied.  See the License for the specific
+ * language governing permissions and limitations under the License.
+ */
+
+package com.radixdlt.api.service.reducer;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Maps;
+import com.radixdlt.atommodel.validators.state.ValidatorMetaData;
+import com.radixdlt.crypto.ECPublicKey;
+import com.radixdlt.identifiers.REAddr;
+import com.radixdlt.statecomputer.ValidatorDetails;
+import com.radixdlt.utils.UInt256;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static com.radixdlt.utils.functional.FunctionalUtils.removeKey;
+import static com.radixdlt.utils.functional.FunctionalUtils.replaceEntry;
+
+import static java.util.Optional.ofNullable;
+
+/**
+ * Wrapper class for registered validators
+ */
+public final class AllValidators {
```

Comment:
Not directly related to your CR but a general question.

I've noticed that many of our classes are modelled as immutable. Do we have general guideline about which class should be immutable?

It seems to me that we're making a lot updates to `AllValidator`?  Is it worth to sacrifice performance  (a lot of copies) for immutability.

cc @talekhinezh

Reply:
Good question. No, I don't think we should stick to immutability for the sake of immutability. Some of these reducer classes are set as immutable because of the way we manage these in our state branching. Changing state and then possibly needing to revert that state was faster to implement with immutable structures. Having said that, I'm definitely looking to removing the immutable reducer stuff soon.

Reply:
I'd prefer to use immutable structures unless I have sensible reasons to have value mutable. Mutable state requires considering possible concurrent execution issues, which creates unnecessary mental overhead.